tools/2025-06-radius-upgrades-flux.md

```diff
@@ -0,0 +1,436 @@
+                       │ HelmRelease     │
+                       │ Resource        │
+                       └─────────────────┘
+                                │
+                                ▼
+                       ┌─────────────────┐
+                       │ Radius          │
+                       │ FluxHelmRelease │
+                       │ Reconciler      │
+                       │                 │
+                       │ • Version Check │
+                       │ • Preflight Run │
+                       │ • Suspend/Resume│
+                       └─────────────────┘
+```
+
+### Component Design
+
+#### FluxHelmReleaseReconciler
+
+**Responsibilities:**
+- Watch Flux HelmRelease resources with opt-in annotation
+- Detect version changes requiring preflight validation
+- Execute preflight checks using existing registry
+- Control HelmRelease progression via suspend/resume
+- Maintain state through annotations
+- Generate observability events
+
+**Key Methods:**
+- `Reconcile()`: Main reconciliation loop
+- `isRadiusChart()`: Validates opt-in annotation
+- `runPreflightChecks()`: Executes validation using existing registry
+- `holdHelmRelease()`: Suspends HelmRelease on preflight failures
+- `clearHoldAndMarkComplete()`: Resumes HelmRelease on success
+
+### State Management
+
+The reconciler uses three key annotations to track state:
+
+1. **`radapp.io/upgrade-enabled: "true"`** (User-controlled)
+   - Opt-in mechanism for safety checks
+   - Must be explicitly added to HelmRelease
+
+2. **`radapp.io/upgrade-checked-version: "0.46.0"`** (Controller-managed)
+   - Tracks last version that passed preflight checks
+   - Prevents redundant check execution
+
+3. **`radapp.io/upgrade-hold: "reason"`** (Controller-managed)
+   - Indicates HelmRelease is suspended due to preflight failure
+   - Contains human-readable failure reason
+
+### Integration Points
+
+#### Preflight Check System
+- Reuses existing `pkg/upgrade/preflight` infrastructure
+- Supports version compatibility checks
+- Extensible for additional validation rules
+- Consistent behavior with CLI upgrade command
+
+#### Flux Integration
+- Uses `spec.suspend` to control HelmRelease progression
+- Leverages Flux's built-in reconciliation patterns
+- Maintains compatibility with Flux GitOps workflows
+
+#### Kubernetes Events
+- Generates standard Kubernetes events for observability
+- Integrates with existing monitoring and alerting systems
+- Provides audit trail for compliance requirements
+
+## API Design
+
+### Annotations
+
+```yaml
+metadata:
+  annotations:
+    # User-controlled: Opt-in to upgrade safety checks
+    radapp.io/upgrade-enabled: "true"
+    
+    # Controller-managed: Last checked version (avoids redundant checks)
+    radapp.io/upgrade-checked-version: "0.46.0"
+    
+    # Controller-managed: Hold reason when upgrades are blocked
+    radapp.io/upgrade-hold: "Cannot upgrade from 0.46.0 to 0.48.0: version skipping not supported"
```

Comment:
What is the rollback scenario here in this case? Will they run helm rollback manually?